File: sdk/typescript/src/audit.ts
/**
 * Audit trail: hash-chain log. Entries are stored locally; only the rolling hash
 * is transmitted to Meridian. This matches spec §13.2: audit logs stay local,
 * hash chain is shared for integrity verification.
 *
 * Log format (one JSON line per entry):
 *   { seq, at, prev_hash, entry_hash, op, meta }
 */
import type { TelemetryEvent } from "./types";

export class AuditChain {
  private prevHash = "GENESIS";
  private seq = 0;
  constructor(private readonly writer: (line: string) => Promise<void> | void) {}

  async append(op: string, meta: Record<string, unknown>): Promise<{ seq: number; entry_hash: string }> {
    this.seq += 1;
    const at = new Date().toISOString();
    const payload = JSON.stringify({ seq: this.seq, at, prev_hash: this.prevHash, op, meta });
    const entry_hash = await sha256(payload);
    const line = JSON.stringify({ seq: this.seq, at, prev_hash: this.prevHash, entry_hash, op, meta });
    await this.writer(line);
    this.prevHash = entry_hash;
    return { seq: this.seq, entry_hash };
  }

  head(): { seq: number; prev_hash: string } {
    return { seq: this.seq, prev_hash: this.prevHash };
  }
}

export async function sha256(s: string): Promise<string> {
  const data = new TextEncoder().encode(s);
  const buf = await crypto.subtle.digest("SHA-256", data);
  return Array.from(new Uint8Array(buf))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

// Default file-backed writer (Node only). Falls back to stdout on other runtimes.
export function defaultWriter(path: string | null): (line: string) => Promise<void> {
  if (!path) return async () => {};
  if (typeof process === "undefined" || typeof require === "undefined") {
    return async (line) => {
      // Browser / edge — append to in-memory ring buffer and stream to stdout if present.
      (globalThis as any).__meridian_audit ??= [];
      (globalThis as any).__meridian_audit.push(line);
    };
  }
  return async (line) => {
    try {
      // Dynamic require to avoid bundler issues in non-Node environments.
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const fs = require("node:fs/promises");
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const pathMod = require("node:path");
      await fs.mkdir(pathMod.dirname(path), { recursive: true });
      await fs.appendFile(path, line + "\n", "utf8");
    } catch {
      /* swallow */
    }
  };
}

File: sdk/python/meridian/audit.py
"""Audit trail: hash-chain log. Entries stored locally; only hash transmitted."""

from __future__ import annotations

import hashlib
import json
import os
import threading
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Callable, Dict


def sha256(data: str) -> str:
    return hashlib.sha256(data.encode("utf-8")).hexdigest()


@dataclass
class AuditHead:
    seq: int
    prev_hash: str


class AuditChain:
    """Append-only hash-chained log. Thread-safe."""

    def __init__(self, path: str | None = ".meridian/audit.log"):
        self._path = path
        self._seq = 0
        self._prev = "GENESIS"
        self._lock = threading.Lock()
        if path:
            Path(os.path.dirname(path) or ".").mkdir(parents=True, exist_ok=True)

    def append(self, op: str, meta: Dict[str, Any]) -> Dict[str, Any]:
        with self._lock:
            self._seq += 1
            at = datetime.now(timezone.utc).isoformat()
            payload = json.dumps(
                {"seq": self._seq, "at": at, "prev_hash": self._prev, "op": op, "meta": meta},
                sort_keys=True,
                separators=(",", ":"),
            )
            entry_hash = sha256(payload)
            line = json.dumps(
                {
                    "seq": self._seq,
                    "at": at,
                    "prev_hash": self._prev,
                    "entry_hash": entry_hash,
                    "op": op,
                    "meta": meta,
                },
                separators=(",", ":"),
            )
            if self._path:
                with open(self._path, "a", encoding="utf-8") as f:
                    f.write(line + "\n")
            self._prev = entry_hash
            return {"seq": self._seq, "entry_hash": entry_hash}

    @property
    def head(self) -> AuditHead:
        return AuditHead(seq=self._seq, prev_hash=self._prev)
